arviz/stats/stats_utils.py

```diff
@@ -0,0 +1,378 @@
+        """General ufunc for single-output function."""
+        if out is None:
+            out = np.empty(ary.shape[:-n_dims])
+        else:
+            if out.shape != ary.shape[:-n_dims]:
+                msg = "Shape incorrect for `out`: {}.".format(out.shape)
+                msg += " Correct shape is {}".format(ary.shape[:-n_dims])
+                raise TypeError(msg)
+        for idx in np.ndindex(out.shape):
+            ary_idx = ary[idx].ravel() if ravel else ary[idx]
+            out[idx] = np.asarray(func(ary_idx, *args, **kwargs))[index]
+        return out
+
+    def _multi_ufunc(ary, *args, out=None, **kwargs):
+        """General ufunc for multi-output function."""
+        element_shape = ary.shape[:-n_dims]
+        if out is None:
+            out = tuple(np.empty(element_shape) for _ in range(n_output))
+        else:
+            raise_error = False
+            correct_shape = tuple(element_shape for _ in range(n_output))
+            if isinstance(out, tuple):
+                out_shape = tuple(item.shape for item in out)
+                if out_shape != correct_shape:
+                    raise_error = True
+            else:
+                raise_error = True
+                out_shape = "not tuple, type={}".format(type(out))
+            if raise_error:
+                msg = "Shapes incorrect for `out`: {}.".format(out_shape)
+                msg += " Correct shapes are {}".format(correct_shape)
+                raise TypeError(msg)
+        for idx in np.ndindex(element_shape):
+            ary_idx = ary[idx].ravel() if ravel else ary[idx]
+            results = func(ary_idx, *args, **kwargs)
+            for i, res in enumerate(results):
+                out[i][idx] = np.asarray(res)[index]
+        return out
+
+    if n_output > 1:
+        ufunc = _multi_ufunc
+    else:
+        ufunc = _ufunc
+
+    update_docstring(ufunc, func, n_output)
+    return ufunc
+
+
+def wrap_xarray_ufunc(
+    ufunc, dataset, *, ufunc_kwargs=None, func_args=None, func_kwargs=None, **kwargs
+):
+    """Wrap make_ufunc with xarray.apply_ufunc.
+
+    Parameters
+    ----------
+    ufunc : callable
+    dataset : xarray.dataset
+    ufunc_kwargs : dict
+        Keyword arguments passed to `make_ufunc`.
+            - 'n_dims', int, by default 2
+            - 'n_output', int, by default 1
+            - 'index', slice, by default Ellipsis
+            - 'ravel', bool, by default True
+    func_args : tuple
+        Arguments passed to 'ufunc'.
+    func_kwargs : dict
+        Keyword arguments passed to 'ufunc'.
+    **kwargs
+        Passed to xarray.apply_ufunc.
+
+    Returns
+    -------
+    xarray.dataset
+    """
+    if ufunc_kwargs is None:
+        ufunc_kwargs = {}
+    if func_args is None:
+        func_args = tuple()
+    if func_kwargs is None:
+        func_kwargs = {}
+
+    n_output = ufunc_kwargs.pop("n_output", 1)
+    callable_ufunc = make_ufunc(ufunc, n_output=n_output, **ufunc_kwargs)

```

Comment:
`ufunc_kwargs.pop("n_output", 1)` is same as: find "n_output" and remove from dict if found, else return 1

This can be other than 1.

`make_ufunc` can be used without `ufunc_wrapper` (not remember what was the name).

Reply:
Yes, what I mean is: if `ufunc_kwargs` has "n_output" as key, `n_output` will be different than one. If "n_output" is not a key in `ufunc_kwargs`, `n_output` will be 1. 
This is the same behavior as passing only `**ufunc_kwargs` to `make_ufunc`: if "n_output" is a key, it will override the default in `make_ufunc`, whereas if it is not a key, the default will be used, which is 1.

Reply:
Oh yes, that is true. I will fix that tomorrow.